src/shadowbox/server/server_access_key.ts

```diff
@@ -198,15 +206,27 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
     }
     const metricsId = uuidv4();
     const password = params?.password ?? generatePassword();
+
     const encryptionMethod = params?.encryptionMethod || this.NEW_USER_ENCRYPTION_METHOD;
+    if (encryptionMethod !== this.NEW_USER_ENCRYPTION_METHOD) {
+      if (!isValidCipher(encryptionMethod)) {
+        throw new errors.InvalidCipher(encryptionMethod);
+      }
+    }
 
-    // Validate encryption method.
-    if (!isValidCipher(encryptionMethod)) {
-      throw new errors.InvalidCipher(encryptionMethod);
+    const portNumber = params?.portNumber ?? this.portForNewAccessKeys;
+    if (portNumber !== this.portForNewAccessKeys) {
+      if (!isValidPort(portNumber)) {
+        throw new errors.InvalidPortNumber(portNumber);
+      }
+      if (!(await this.isPortAvailable(portNumber))) {
+        throw new errors.PortUnavailable(portNumber);
+      }
     }
```

Comment:
Your portNumber assignment is not a clean assignment, because portNumber can be anything. This makes `portNumber` essentially `any`.

The `portNumber !== this.portForNewAccessKeys` comparison is comparing different types, which is also not ideal.

You should only assign `params?.portNumber` after you validated it.

```suggestion
    var portNumber = this.portForNewAccessKeys;
    if (params?.portNumber !== undefined) {
      if (!isValidPort(params?.portNumber)) {
        throw new errors.InvalidPortNumber(params?.portNumber);
      }
      portNumber = params?.portNumber;
      if (!(await this.isPortAvailable(portNumber))) {
        throw new errors.PortUnavailable(portNumber);
      }
    }
```

This relies on changing the `isValidPort` to use a type predicate. That tells the compiler that `params?.portNumber` is actually a number after the validity test.

Reply:
I think you may be confusing the types between the `ShadowsocksManagerService` (where we need to validate and type narrow the provided `port` parameter) and the `ServerAccessKeyRepository` here (where the `portNumber` is always a number).  Take a look at the `AccessKeyCreateParams` interface in `access_key.ts` where we add `portNumber`.

I can validate before assignment if you prefer (that's probably just a style preference difference), but I just want to clarify there is no type benefit to that at all.

Reply:
You are right, I was confusing it with the service! Nevermind.